### PR TITLE
feat(scheduler): scheduler change after node leave

### PR DESCRIPTION
**Pull request by JasonOE**
# Refactor Pipeline Validation and Improve Node State Management

## Summary

This PR refactors the pipeline validation logic, improves node state management, and enhances error handling in the scheduler. The main changes include:

1. **Unified pipeline validation**: Consolidates `has_full_active_pipeline()` into `has_full_pipeline()` with an `active_only` parameter
2. **Improved pipeline detection algorithm**: Replaces simple layer counting with a DFS-based approach that correctly validates contiguous pipeline chains
3. **Better node state management**: Ensures `is_active` flag is properly set when nodes are deallocated
4. **Enhanced error handling**: Adds proper logging and state updates for global rebalancing operations
5. **Code cleanup**: Removes unnecessary debug logging statements

#### 1. Unified Pipeline Validation (`layer_allocation.py`)

- **Before**: Two separate methods `has_full_pipeline()` and `has_full_active_pipeline()` with different implementations
- **After**: Single `has_full_pipeline(active_only: bool = False)` method that handles both cases
- **Algorithm improvement**: Replaced simple layer counting with DFS-based validation that correctly checks for contiguous node chains from layer 0 to `num_total_layers`

The new implementation:
- Builds an index of nodes by `start_layer`
- Uses DFS to verify a continuous path exists from layer 0 to the end
- Properly filters inactive nodes when `active_only=True`

#### 2. Node State Management (`layer_allocation.py`)

- Added `node.is_active = False` in `deallocate()` to ensure nodes are marked inactive when deallocated
- This prevents inactive nodes from being considered in pipeline validation

#### 3. Enhanced Error Handling (`scheduler.py`)

- Added proper success/failure logging for global rebalancing operations
- Ensures `_bootstrapped` flag and `_bootstrapped_event` are correctly updated after rebalancing
- Provides better visibility into scheduler state transitions

#### 4. API Updates (`scheduler_manage.py`)

- Updated `get_schedule_status()` to use the new unified `has_full_pipeline(active_only=True)` method
- Maintains the same functionality while using the improved implementation

### Code Cleanup

- Removed debug logging statements from:
  - `model_info.py`: Removed verbose logging in `decoder_layer_io_bytes()`
  - `node.py`: Removed debug logging in `get_decoder_layer_capacity()`

## Benefits

1. **Correctness**: The DFS-based pipeline validation correctly identifies contiguous pipelines, fixing potential edge cases with the previous layer-counting approach
2. **Maintainability**: Unified API reduces code duplication and makes the codebase easier to maintain
3. **Reliability**: Proper node state management ensures inactive nodes don't interfere with pipeline validation
4. **Observability**: Better logging helps diagnose scheduler state issues

## Testing

- Existing tests should continue to pass with the improved pipeline validation logic
- The DFS-based approach is more robust and handles edge cases better than the previous implementation

## Migration Notes

- No breaking changes for external APIs
- Internal callers of `has_full_active_pipeline()` should use `has_full_pipeline(active_only=True)` instead

### CODE REVIEW

**Review by sl-gn — APPROVED** [review]

File: src/scheduling/layer_allocation.py [hunk 1]
```diff
@@ -529,39 +529,64 @@ def _adjust_end_layer_for_tail(self, node: Node, proposed_start_layer: int) -> i
...
+                return current_end == total_layers
+
+            for nxt in start_to_nodes.get(current_end, []):
+                if nxt.end_layer and nxt.end_layer > current_end:
+                    if can_reach_target(nxt.end_layer):
+                        return True
+            return False
+
+        return any(
+            head.end_layer and can_reach_target(head.end_layer)
```
Comment (christ-tt) [inline src/scheduling/layer_allocation.py]: Why we need this check? Say we have 3 nodes, 2 active and 1 idling, then the idling one will be assigned 'lightest' layers (measured by load), which is not necessarily connected all the way to the `end_layer`. Suppose we have total 48 layers, and 3 identical nodes each can host 24 layers, then we will have ([0-24) [24-48)] and the last node, now idling, will handle [0-24).
By our allocation design, we want any newly join nodes connect the idling ones and fill the gap of `layer_loads`.

Add unit tests to guard your changes don't break previous design.

**Review by christ-tt — APPROVED** [review]
LGTM after minor fix. Good work and thanks a lot!

File: src/scheduling/model_info.py [hunk 2]
```diff
@@ -184,12 +184,12 @@ def decoder_layer_io_bytes(
                 ffn_params *= self.num_local_experts
             kv_cache_size = 0
 
-        logger.debug(
-            "Model Info ffn_params=%d, kv_cache_size=%d, attention_params=%d",
-            ffn_params,
-            kv_cache_size,
-            attention_params,
-        )
+        # logger.debug(
```
Comment (gufengc) [inline src/scheduling/model_info.py]: why?

**Comment by christ-tt** [comment]
I think we can merge this for now. 
As a TODO, let's consider this case: suppose we have 3 identical nodes, our model has 28 layers, each node can host 22 layers. After bootstrapping, we will have node 1: `[0, 14)`, node 2: `[14, 28)`, and node 3, currently idling, hosting `[0, 22)`. Now, suppose node 1 left, we now have node 2 and node 3; here we don't necessarily need **global** rebalance - node 2 can stay unchanged, and we only need to trim the layer hosted on node 3 ([0, 22) -> [0, 14)). 
We have very similar [logic](https://github.com/GradientHQ/parallax/blob/fd508211be3b8b28dab7485be27a8d7e19e1b74e/src/scheduling/request_routing.py#L33) implemented in request routing phase. I think we can try to implement a separate function here in `layer_allocation`, checking if **trimming** any nodes makes sense, and put this **before** `should_global_rebalance` check. Especially for `active` but **idling** (not forming any actual pipeline, or serving any request) nodes.
But for now, I think current implementation is good enough. We can open another PR handling this optimization.
Thanks.
